src/test/bpf_query.c

```diff
@@ -0,0 +1,107 @@
+/* -*- Mode: C; tab-width: 8; c-basic-offset: 2; indent-tabs-mode: nil; -*- */
+#include "util.h"
+
+#include <stdint.h>
+#include <linux/bpf.h>
+
+#define TEST_CGROUP_NAME "rr_test_cgroup"
+#define TEST_CGROUP_PATH "/sys/fs/cgroup/" TEST_CGROUP_NAME
+#define MAX_PROG_CNT 2
+
+int bpf(int cmd, union bpf_attr *attr, unsigned int size) {
+  return syscall(__NR_bpf, cmd, attr, size);
+}
+
+const struct bpf_insn bpf_program[] = {
+  // mov r0, 0
+  { .code = BPF_ALU | BPF_MOV | BPF_K, .dst_reg = 0, .imm = 0 },
+  // exit
+  { .code = BPF_JMP | BPF_EXIT },
+};
+
+
+int main(void) {
+  __u32 prog_ids[MAX_PROG_CNT];
+
+  union bpf_attr query_attr = {
+    .query = {
+      .prog_ids = (uintptr_t)&prog_ids,
+    },
+  };
+
+  // cleanup the test cgroup, if any
+  if (rmdir(TEST_CGROUP_PATH) != 0) {
+    test_assert(errno != ENOTEMPTY);
+  }
+
+  // create the test cgroup
+  if (mkdir(TEST_CGROUP_PATH, 0755) != 0) {
+    if (errno == EACCES) {
+      atomic_puts("Skipping test because it requires CAP_SYS_ADMIN");
+      atomic_puts("EXIT-SUCCESS");
+      return 0;
+    }
+    if (errno == ENOENT) {
+      atomic_puts("No cgroup vfs mounted on /sys/fs/cgroup, skipping");
+      atomic_puts("EXIT-SUCCESS");
+      return 0;
+    }
+    test_assert(!"failed to create cgroup");
```

Comment:
So how does this test work when you run multiple copies simultaneously? Can we create a unique cgroup per test?

Reply:
Or can we use socket or perf eBPF programs instead so it's simpler to set up the right environment?

Reply:
Or are cgroups the only way to test `BPF_PROG_ATTACH`?

Reply:
> So how does this test work when you run multiple copies simultaneously? Can we create a unique cgroup per test?

I'll create a cgroup per test

> Or can we use socket or perf eBPF programs instead so it's simpler to set up the right environment?

I believe socket / perf programs use another API. Here the full list of attach types:
https://www.kernel.org/doc/html/latest/bpf/libbpf/program_types.html

Most are cgroup stuff and kernel probes. I wish one of the types could be attached to an easier to create fd, but can't find any. I believe setting up a kprobe would be slightly easier, but would also have concurrency issues.